daisy_workflows/image_import/suse/translate.py

```diff
@@ -26,67 +26,159 @@
+        subscriptions=['sle-module-public-cloud/12/x86_64']
+    ),
+]
+
+
+def get_distro(g) -> SuseRelease:
+  for d in _distros:
+    if d.flavor == g.gcp_image_distro:
+      if d.major == g.gcp_image_major or d.major == '*':
+        if d.minor == g.gcp_image_minor or d.minor == '*':
+          return d
+  raise AssertionError('Import script not defined for {} {}.{}'.format(
+      g.gcp_image_distro, g.gcp_image_major, g.gcp_image_minor))
+
+
+def install_subscriptions(distro, g):
+  if distro.subscriptions:
+    for subscription in distro.subscriptions:
+      try:
+        g.command(['SUSEConnect', '-p', subscription])
+      except Exception as e:
+        raise ValueError(
+            'Command failed: SUSEConnect -p {}: {}'.format(subscription, e))
+
+
+def install_packages(distro, g, install_gce):
+  g.command(['zypper', 'refresh'])
+  for pkg in distro.packages:
+    if pkg.gce and not install_gce:
+      continue
+    try:
+      g.command(('zypper', '-n', 'install', '--no-recommends', pkg.name))
+    except Exception as e:
+      if pkg.required:
+        raise AssertionError(
+            'Failed to install required package {}: {}'.format(pkg.name, e))
+      else:
+        logging.warning(
+            'Failed to install optional package {}: {}'.format(pkg.name, e))
 
 
-def DistroSpecific(g):
-  install_gce = utils.GetMetadataAttribute('install_gce_packages')
+def update_grub(g):
+  g.command(
+      ['sed', '-i',
+       r's#^\(GRUB_CMDLINE_LINUX=".*\)"$#\1 console=ttyS0,38400n8"#',
+       '/etc/default/grub'])
+  g.command(['grub2-mkconfig', '-o', '/boot/grub2/grub.cfg'])
 
-  # Remove any hard coded DNS settings in resolvconf.
-  logging.info('Resetting resolvconf base.')
-  g.sh('echo "" > /etc/resolv.conf')
 
-  # Try to reset the network to DHCP.
-  g.write('/etc/sysconfig/network/ifcfg-eth0', network)
+def reset_network(g):
+  logging.info('Updating network to use DHCP.')
+  g.sh('echo "" > /etc/resolv.conf')
+  g.write('/etc/sysconfig/network/ifcfg-eth0', '\n'.join((
+      'BOOTPROTO=dhcp',
+      'STARTMODE=auto',
+      'DHCLIENT_SET_HOSTNAME=yes')
+  ))
 
-  if install_gce == 'true':
-    g.command(['zypper', 'refresh'])
 
-    logging.info('Installing cloud-init.')
-    g.command(['zypper', '-n', 'install', '--no-recommends', 'cloud-init'])
+def install_gcp_licensing(distro, g):
+  if distro.flavor == 'opensuse':
+    return
+  licensing = utils.GetMetadataAttribute('licensing')
+  if not licensing:
+    raise AssertionError(
+        'licensing attribute must be set to either gcp or byol.')
+  if licensing == 'byol':
+    return
 
-    # Installing google-compute-engine-init and not installing
-    # gce-compute-image-packages as there is no port to SUSE
-    logging.info('Installing GCE packages.')
-    g.command(
-        ['zypper', '-n', 'install', '--no-recommends',
-        'google-compute-engine-init'])
+  # TODO: Implement GCP licensing for SLES.
```

Comment:
How are you tracking this work? Can this script receive 'gcp' licensing as the code is currently set up?

Reply:
To simplify this review, I went ahead and removed this method